### PR TITLE
Add source locations to syntax test expectations.

### LINKED ISSUE #3810: Include source code locations in syntax test expectations.
Include and check source code locations in the expectations defined in the test contracts.

Came up in #3808.

### PR DESCRIPTION
**Pull request by ekpyron**
Also uses colored background to indicate warnings and errors when printing the contract in isoltest.

Closes #3810.

### CODE REVIEW

File: test/tools/isoltest.cpp [hunk 1]
```diff
@@ -90,11 +89,52 @@ string SyntaxTestTool::editor;
...
 {
-	stringstream stream(m_test->source());
-	string line;
-	while (getline(stream, line))
-		cout << "    " << line << endl;
-	cout << endl;
+	if (m_formatted)
+	{
+		std::string const& source = m_test->source();
+		if (source.empty()) return;
```
Comment (erak) [inline test/tools/isoltest.cpp]: I'd prefer moving ``return`` to a new line.

File: test/libsolidity/SyntaxTest.cpp [hunk 2]
```diff
@@ -47,6 +47,27 @@ void skipSlashes(IteratorType& it, IteratorType end)
 		++it;
 }
 
+void expect(string::iterator& it, string::iterator end, string::value_type c)
```
Comment (erak) [inline test/libsolidity/SyntaxTest.cpp]: Please use the ``_`` prefix for function arguments.

**Review by erak — DISMISSED** [review]
Besides the tiny style changes this looks good.

File: test/libsolidity/SyntaxTest.cpp [hunk 3]
```diff
@@ -34,17 +34,38 @@ namespace fs = boost::filesystem;
...
+}
+
+void expect(string::iterator& _it, string::iterator _end, string::value_type _c)
+{
+	if (_it == _end || *_it != _c)
+		throw runtime_error(string("Invalid test expectation. Expected: \"") + _c + "\".");
+	++_it;
+}
+
+int parseLocation(string::iterator& _it, string::iterator _end)
```
Comment (axic) [inline test/libsolidity/SyntaxTest.cpp]: This isn't actually any specific about location, it just parses an unsigned number.

File: test/tools/isoltest.cpp [hunk 4]
```diff
@@ -262,10 +299,6 @@ SyntaxTestStats SyntaxTestTool::processPath(
 				paths.pop();
 				++successCount;
 				break;
-			default:
```
Comment (axic) [inline test/tools/isoltest.cpp:265]: Do we have a compiler flag which will result in an error if we extend the `enum` but not this `switch`?

File: test/libsolidity/SyntaxTest.cpp [hunk 5]
```diff
@@ -60,22 +81,39 @@ SyntaxTest::SyntaxTest(string const& _filename)
...
 	if (m_compiler.parse())
 		m_compiler.analyze();
 
 	for (auto const& currentError: filterErrors(m_compiler.errors(), true))
-		m_errorList.emplace_back(SyntaxTestError{currentError->typeName(), errorMessage(*currentError)});
+	{
+		int locationStart = -1, locationEnd = -1;
+		if (auto location = boost::get_error_info<errinfo_sourceLocation>(*currentError))
+		{
+			if (location->start >= versionPragmaLength)
```
Comment (axic) [inline test/libsolidity/SyntaxTest.cpp]: Perhaps a comment here could help: ignore the version pragma inserted by the testing tool when calculating locations.

File: test/tools/isoltest.cpp (hunk 4 above)
Comment (chriseth) [reply test/tools/isoltest.cpp:265]: This is usually the case at least on my machine.

File: test/libsolidity/SyntaxTest.cpp [hunk 6]
```diff
@@ -60,22 +81,39 @@ SyntaxTest::SyntaxTest(string const& _filename)
 
 bool SyntaxTest::run(ostream& _stream, string const& _linePrefix, bool const _formatted)
 {
+	char const versionPragma[] = "pragma solidity >=0.0;\n";
```
Comment (chriseth) [inline test/libsolidity/SyntaxTest.cpp]: Why not use `string` here?

Comment (chriseth) [reply test/libsolidity/SyntaxTest.cpp]: This would also remove the need for the length variable.

**Review by chriseth — DISMISSED** [review]

File: test/tools/isoltest.cpp (hunk 4 above)
Comment (ekpyron) [reply test/tools/isoltest.cpp:265]: In this particular case even it shouldn't be a problem even if there is no compiler error - the only reason for extending this particular enum would be to distinguish cases in exactly this switch.

File: test/tools/isoltest.cpp [hunk 7]
```diff
@@ -90,11 +89,53 @@ string SyntaxTestTool::editor;
...
 void SyntaxTestTool::printContract() const
 {
-	stringstream stream(m_test->source());
-	string line;
-	while (getline(stream, line))
-		cout << "    " << line << endl;
-	cout << endl;
+	if (m_formatted)
+	{
+		std::string const& source = m_test->source();
```
Comment (axic) [inline test/tools/isoltest.cpp]: `std` should not needed in this file.

File: test/tools/isoltest.cpp [hunk 8]
```diff
@@ -90,11 +89,53 @@ string SyntaxTestTool::editor;
...
+		if (source.empty())
+			return;
+
+		std::vector<char const*> sourceFormatting(source.length(), formatting::RESET);
+		for (auto const& error: m_test->errorList())
+			if (error.locationStart >= 0 && error.locationEnd >= 0)
+			{
+				solAssert(static_cast<size_t>(error.locationStart) < source.length(), "");
```
Comment (axic) [inline test/tools/isoltest.cpp]: I wonder if we should use `solAssert` here or just `assert`?

Comment (axic) [reply test/tools/isoltest.cpp]: I think it shouldn't be `solAssert` as this method is called outside where Solidity specific exceptions are handled.

Comment (chriseth) [reply test/tools/isoltest.cpp]: I think we should promote solAssert to be not specific about solidity anymore.

Comment (ekpyron) [reply test/tools/isoltest.cpp]: Wouldn't the not-solidity-specific version of ``solAssert`` be libdevcore's ``assertThrow``?
``solAssert`` throws ``dev::solidity::InternalCompilerError`` which is rather solidity-specific itself.

For the moment I changed to plain ``assert`` in ``isoltest.cpp``.

**Comment by ekpyron** [comment]
Before merging this we should make sure that it is rebased to develop to make sure that no new syntax tests have been introduced by other merges that will fail if not updated to include source locations.

**Review by axic — APPROVED** [review]

**Pull request merged**